extensions/amp-form/0.1/form-verifiers.js

```diff
@@ -133,30 +62,85 @@ function isSubmittable_(element) {
 export class FormVerifier {
   /**
    * @param {!HTMLFormElement} form
-   * @param {!Array<!VerificationGroup>} groups
    */
-  constructor(form, groups) {
+  constructor(form) {
     /** @protected @const */
     this.form_ = form;
+  }
 
-    /** @protected @const {!Array<!VerificationGroup>} */
-    this.groups_ = groups;
+  /**
+   * Called when the user has fully set a value to be verified,
+   * e.g. the input's 'change' event
+   * @return {!Promise<!Array<!Element>>}
+   */
+  onCommit() {
+    this.clearVerificationErrors_();
+    if (this.isDirty_()) {
+      return this.verify_();
+    } else {
+      return Promise.resolve([]);
+    }
   }
 
   /**
-   * Called when the user has modified the value in the input,
-   * e.g. the input's 'input' event
-   * @param {!Element} unusedInput
+   * Sends the verify request if any group is ready to verify.
+   * @return {!Promise<!Array<!Element>>} The list of elements whose state
+   *    must change
+   * @protected
    */
-  onMutate(unusedInput) {}
+  verify_() {
+    return Promise.resolve([]);
+  }
 
   /**
-   * Called when the user has fully set a value to be verified,
-   * e.g. the input's 'change' event
-   * @param {!Element} unusedInput
-   * @param {!function(!Array<!Element>)} unusedAfterVerify
+   * Checks if the form has been changed from its initial state.
+   * @private
+   */
+  isDirty_() {
+    const form = this.form_;
+    for (let i = 0; i < form.elements.length; i++) {
+      const field = form.elements[i];
```

Comment:
Ah I didn't consider it'd be computed every access. I'll cache it in a variable outside the loop.